tiledb/sm/array/consistency.cc

```diff
@@ -0,0 +1,113 @@
+/**
+ * @file   consistency.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2017-2022 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements classes ConsistencyController and ConsistencySentry.
+ */
+
+#include <iostream>
+
+#include "tiledb/sm/array/consistency.h"
+
+using namespace tiledb::common;
+
+namespace tiledb::sm {
+
+ConsistencySentry::ConsistencySentry(
+    ConsistencyController& registry, ConsistencyController::entry_type entry)
+    : parent_(registry)
+    , entry_(entry) {
+}
+
+ConsistencySentry::ConsistencySentry(ConsistencySentry&& x)
```

Comment:
Where is this used?

Reply:
this is used in `Array::set_array_open()`. With these changes, the array [de]registration has been moved into the `set_array_[open/closed]` functions. An _optional_ Sentry object is created in the Array's initializer list so _temporary_ copies are made by calling `make_sentry`. Thus, a copy constructor had to be added to properly get the just-created Sentry into the Array's member variable in `set_array_open()`.

Reply:
Thanks!

Reply:
If I am not mistaken though this is actually a `move` constructor, no?
Can you add a comment that explain why this line is needed `x.entry_ = parent_.array_registry_.end();` ?
I think it's because we want `x` to be properly freed by the distructor after, but I'll let you confirm.

Reply:
Sorry, @ypatia, I misspoke above. You are correct - this is a _move_ constructor, NOT a _copy_ constructor. 

> I think it's because we want `x` to be properly freed by the distructor after
_Sort of_. It's really transferring the iterator from the argument object to the new object. `std::multimap.end()` returns the an iterator referring to the past-the-last-element. It doesn't actually point to anything and isn't dereferenceable. So, this function call is more of a pseudo-delete on the argument's iterator.